src/fmmax/utils.py

```diff
@@ -142,27 +145,57 @@ def _eig_cpu(matrix: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
     eigenvalues, eigenvectors = _eig_host(matrix)
-    return (eigenvalues, eigenvectors), (eigenvalues, eigenvectors, eps)
+    return (eigenvalues, eigenvectors), (eigenvalues, eigenvectors, eps_relative)
 
 
 def _eig_bwd(
     res: Tuple[jnp.ndarray, jnp.ndarray, float],
     grads: Tuple[jnp.ndarray, jnp.ndarray],
 ) -> Tuple[jnp.ndarray, None]:
     """Implements the backward calculation for `eig`."""
-    eigenvalues, eigenvectors, eps = res
+    eigenvalues, eigenvectors, eps_relative = res
     grad_eigenvalues, grad_eigenvectors = grads
 
-    # Compute the broadened F-matrix. The expression is similar to that of equation 5
-    # from [2021 Colburn], but differs slightly from both the code and paper.
-    # TODO: derive the proper expression for broadened F-matrix.
+    # The expression for gradient of matrix eigenvectors with respect to matrix values
+    # contains a difference between eigenvalues in the denominator. This causes
+    # numerical problems in situations where eigenvalues are degenerate or nearly
+    # degenerate. This problem is addressed in a few different ways in existing rcwa
+    # codes. Defining `delta = eigval_i - eigval_j`,
+    #
+    # - torcwa uses Lorentzian broadening of the form
+    #      1 / delta -> delta.conj / (abs(delta)**2 + eps)
+    #   with an eps value of 1e-10.
+    #   (https://github.com/kch3782/torcwa/blob/main/torcwa/torch_eig.py#L29)
+    #
+    # - tf_rcwa uses an expression similar to torcwa, but with apparently different
+    #   conjugation. Their value of eps is 1e-6. (We tested their implementation and
+    #   found it fails some of our gradient validation tests.)
+    #   (https://github.com/scolburn54/rcwa_tf/blob/master/src/tensor_utils.py#L89)
+    #
+    # - grcwa uses `1 / delta -> 1 / (delta + eps)` with an eps value of 1e-10.
+    #   (https://github.com/weiliangjinca/grcwa/blob/master/grcwa/primitives.py#L44)
+    #
+    # Also, from the wider literature:
+    #
+    # - Liao et al. (https://arxiv.org/pdf/1903.09650 sec III.A.1) uses Lorentzian
+    #   broadening with an eps value of 1e-12.
+    #
+    # One issue with Lorentzian broadening and a fixed eps is the independence from
+    # matrix scale. While a given eps may be appropriate for a matrix, it would be
+    # inappropriate for the same matrix scaled by e.g. 1e-6. (The gradient should
+    # be identical, except scaled by 1 / 1e-6.)
+    #
+    # Therefore, we use Lorentzian broadening similar to torcwa, but with an eps
+    # value that is computed in a way that considers the eigenvalue range.
     eigenvalues_i = eigenvalues[..., jnp.newaxis, :]
     eigenvalues_j = eigenvalues[..., :, jnp.newaxis]
     delta_eig = eigenvalues_i - eigenvalues_j
+    eig_range = jnp.amax(jnp.abs(delta_eig), axis=(-2, -1), keepdims=True)
+    eps = jnp.maximum(eps_relative * eig_range, EIG_EPS_MINIMUM)
     f_broadened = delta_eig.conj() / (jnp.abs(delta_eig) ** 2 + eps)
```

Comment:
If the aim is to make the regularization scheme invariant to a global scale factor of the matrix, I think `eps` should be made proportional to the *square* of the absolute value of `delta_eig` instead, no?

Reply:
Ah yes, good catch. I just adjusted the calculation to have eps depend on the square of the maximum eigenvalue difference. Tests still pass so I haven't touched the `EIG_EPS` module constant.